### PR TITLE
go: sql nodes for gorm framework

### PR DESCRIPTION
**Pull request by hiteshbedre**
Description

We are building graph for SQL nodes. In Gorm framework a module defined is act as table in sql and its member as column. So we are capturing TypeDecl nodes on whom gorm related functions are written.
Ex.

      import (
		"github.com/jinzhu/gorm"
	)

	type User struct {
		// we need to use our own custom model because the id is different of the "gorm.Model"
		ID      string `gorm:"PRIMARY KEY; UNIQUE" json:"id"`
		Name    string `gorm:"type:varchar(255); NOT NULL" json:"name" validate:"required"`
		Email   string `gorm:"type:varchar(255)" json:"email"`
		Phone   string `gorm:"type:varchar(100); NOT NULL; UNIQUE; UNIQUE_INDEX" json:"phone" validate:"required"`
		Address string `gorm:"type:text" json:"address"`
		MyModel
	}

	// Save saves a User inside the database
	func (u *User) Save(db *gorm.DB) error {
		uuidResult, err := uuid.NewRandom()
		if err != nil {
			log.Fatal(err)
		}

		u.ID = uuidResult.String()
		u.CreatedAt = time.Now().String()
		u.UpdatedAt = time.Now().String()

		err = db.Save(&u).Error
		if err != nil {
			return err
		}
		return nil

In above code snippet we are capturing method node having parameter `gorm.DB` and processing its corresponding `TypeDecl` node to create sql related nodes.

### CODE REVIEW

File: src/main/scala/ai/privado/languageEngine/go/passes/orm/GormParser.scala [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+
+import scala.collection.mutable
+import scala.util.{Failure, Success, Try}
+
+class GormParser(cpg: Cpg) extends PrivadoParallelCpgPass[TypeDecl](cpg) {
+  val GORM_PARAMETER_TYPE_RULE = ".*github.com/jinzhu/gorm.*"
+  val logger                   = LoggerFactory.getLogger(getClass)
+
+  override def generateParts(): Array[_ <: AnyRef] = {
+    cpg.method.where(_.parameter.typeFullName(GORM_PARAMETER_TYPE_RULE)).typeDecl.dedup.toArray
```
Comment (khemrajrathore) [inline src/main/scala/ai/privado/languageEngine/go/passes/orm/GormParser.scala:47]: change query to `cpg.typeDecl.where(_.method.parameter...)` as there will be always less typeDecl node as compared to method nodes

File: src/main/scala/ai/privado/languageEngine/go/passes/orm/GormParser.scala [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+      try {
+        val sqlTable: SQLTable = SQLTable(
+          model.code,
+          model.lineNumber.getOrElse(Integer.valueOf(-1)),
+          model.columnNumber.getOrElse(Integer.valueOf(-1))
+        )
+        val sqlColumns: List[SQLColumn] = model.member.l.map(x =>
+          SQLColumn(x.code, x.lineNumber.getOrElse(Integer.valueOf(-1)), x.columnNumber.getOrElse(Integer.valueOf(-1)))
+        )
+        val queryModel = SQLQuery(SQLQueryType.CREATE, sqlTable, sqlColumns)
```
Comment (khemrajrathore) [inline src/main/scala/ai/privado/languageEngine/go/passes/orm/GormParser.scala:75]: will this be always be create Query?

Comment (khemrajrathore) [reply src/main/scala/ai/privado/languageEngine/go/passes/orm/GormParser.scala:75]: or as this is a typeDecl node this is as good as Create query

File: src/main/scala/ai/privado/languageEngine/go/passes/orm/GormParser.scala [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+          model.code,
+          model.lineNumber.getOrElse(Integer.valueOf(-1)),
+          0
+        )
+
+      } catch {
+        case ex: Exception =>
+          ex.printStackTrace()
+          println(s"Error while building sql nodes for GORM framework: ${ex.getMessage}")
+          None
```
Comment (khemrajrathore) [inline src/main/scala/ai/privado/languageEngine/go/passes/orm/GormParser.scala:89]: As the function returns Unit, we don't need to specify `None`

File: src/test/scala/ai/privado/languageEngine/go/passes/orm/GormParserTest.scala [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+    super.beforeAll()
+    X2Cpg.applyDefaultOverlays(cpg)
+
+    val context = new LayerCreatorContext(cpg)
+    val options = new OssDataFlowOptions()
+    new OssDataFlow(options).run(context)
+    new GormParser(cpg).createAndApply()
+  }
+
+  override val goFileContents =
```
Comment (khemrajrathore) [inline src/test/scala/ai/privado/languageEngine/go/passes/orm/GormParserTest.scala:45]: In this test Case, which method gets matched with `cpg.method.where(_.parameter.typeFullName(GORM_PARAMETER_TYPE_RULE)).typeDecl.dedup.toArray` this query. ?
As I understand the Ty peDecl node we get will be of `User`

File: src/main/scala/ai/privado/languageEngine/go/passes/orm/GormParser.scala (hunk 2 above)
Comment (hiteshbedre) [reply src/main/scala/ai/privado/languageEngine/go/passes/orm/GormParser.scala:75]: A model is reflection of table schema. In terms sql terminology create just suits better.

File: src/test/scala/ai/privado/languageEngine/go/passes/orm/GormParserTest.scala (hunk 4 above)
Comment (hiteshbedre) [reply src/test/scala/ai/privado/languageEngine/go/passes/orm/GormParserTest.scala:45]: `Save` method will get matched to query. And `Save` is part of TypeDecl `User`.

File: src/main/scala/ai/privado/languageEngine/go/passes/orm/GormParser.scala (hunk 1 above)
Comment (hiteshbedre) [reply src/main/scala/ai/privado/languageEngine/go/passes/orm/GormParser.scala]: Agree, addressed here: 245840ffc166800c843d782ae2efcc5abbb5d02f

File: src/main/scala/ai/privado/languageEngine/go/passes/orm/GormParser.scala (hunk 3 above)
Comment (hiteshbedre) [reply src/main/scala/ai/privado/languageEngine/go/passes/orm/GormParser.scala]: Addressed here: 245840ffc166800c843d782ae2efcc5abbb5d02f

File: src/test/scala/ai/privado/languageEngine/go/passes/orm/GormParserTest.scala (hunk 4 above)
Comment (khemrajrathore) [reply src/test/scala/ai/privado/languageEngine/go/passes/orm/GormParserTest.scala:45]: Ok, looks like this is handled as part of AST creation differently. 

My only concern is False positives due to this, I hope you have already tested this across different cases, if YES that should be good

Comment (hiteshbedre) [reply src/test/scala/ai/privado/languageEngine/go/passes/orm/GormParserTest.scala:45]: Just to simplify purpose of query "Get all TypeDecl node which are dealing with GORM dependency".
TypeDecl node are only created when they actually defined in code base. As parse is tightly coupled with GORM I don't see any case where false positive may arise.

**Review by khemrajrathore — DISMISSED** [review]
LGTM

File: src/main/scala/ai/privado/languageEngine/go/passes/orm/GormParser.scala [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+        buildAndAddSqlQueryNodes(builder, model, fileNode)
+    }
+  }
+
+  private def buildAndAddSqlQueryNodes(
+    builder: DiffGraphBuilder,
+    model: TypeDecl,
+    fileNode: NodeOrDetachedNode
+  ): Unit = {
+    Try {
```
Comment (pandurangpatil) [inline src/main/scala/ai/privado/languageEngine/go/passes/orm/GormParser.scala:65]: Why there are two consicutive `Try` blocks ?

Comment (hiteshbedre) [reply src/main/scala/ai/privado/languageEngine/go/passes/orm/GormParser.scala]: First `Try` was to return Success/Failure object but we are not processing return object so removed `Try` block. `try` block is persisted.

**Comment by khemrajrathore** [comment]
@hiteshbedre Why this PR was closed?

**Pull request closed without merging**